src/e2e_tests.rs

```diff
@@ -0,0 +1,649 @@
+use super::http_api::HttpApi;
+use super::http_workers::HttpWorker;
+use super::identity;
+use super::identity::{Identity, Signer};
+use super::processor;
+use super::proxy::ProxyWorker;
+use super::redis;
+use super::storage::{ProxyStorage, RedisStorage, Storage};
+use super::twin::{Twin, TwinDB};
+use super::types::Message;
+use anyhow::{Context, Result};
+use bb8_redis::redis::AsyncCommands;
+use sp_core::crypto::Pair;
+use sp_core::{ed25519::Pair as EdPair, sr25519::Pair as SrPair};
+use std::collections::HashMap;
+use std::panic;
+use std::process::Command;
+use std::sync::Once;
+use tokio::time::{sleep, Duration};
+
+#[derive(Default, Clone)]
+struct InMemoryDB {
+    pub twins: HashMap<u32, Twin>,
+}
+
+impl InMemoryDB {
+    fn add(&mut self, twin: Twin) {
+        self.twins.insert(twin.id, twin);
+    }
+}
+
+#[async_trait::async_trait]
+impl TwinDB for InMemoryDB {
+    async fn get_twin(&self, twin_id: u32) -> anyhow::Result<Option<Twin>> {
+        Ok(self.twins.get(&twin_id).map(|t| t.clone()))
+    }
+
+    async fn get_twin_with_account(
+        &self,
+        _account_id: sp_runtime::AccountId32,
+    ) -> anyhow::Result<u32> {
+        unimplemented!()
+    }
+}
+
+fn new_message(
+    command: &str,
+    expiration: u64,
+    retry: usize,
+    data: &str,
+    destination: Vec<u32>,
+) -> Message {
+    let ret_queue = format!("{}", uuid::Uuid::new_v4());
+    let epoch = std::time::SystemTime::now()
+        .duration_since(std::time::SystemTime::UNIX_EPOCH)
+        .unwrap()
+        .as_secs();
+    Message {
+        command: String::from(command),
+        expiration: expiration,
+        retry: retry,
+        data: data.to_string(),
+        destination: destination,
+        reply: ret_queue,
+        timestamp: epoch,
+        ..Message::default()
+    }
+}
+
+async fn send_all(messages: Vec<Message>, local_redis: &str) -> Result<(usize, Vec<String>)> {
+    let pool = redis::pool(local_redis)
+        .await
+        .context("unable to create redis connection")?;
+    let mut conn = pool.get().await.context("unable to get redis connection")?;
+    let queue = "msgbus.system.local";
+    let mut responses_expected = 0;
+    let mut return_queues = Vec::new();
+    for msg in messages {
+        let _ = conn.lpush(&queue, &msg).await?;
+        responses_expected += msg.destination.len();
+        return_queues.push(msg.reply);
+    }
+    Ok((responses_expected, return_queues))
+}
+
+async fn wait_for_responses(
+    responses_expected: usize,
+    return_queues: Vec<String>,
+    timeout: usize,
+    local_redis: &str,
+) -> Result<(Vec<Message>, usize, usize, usize)> {
```

Comment:
This is confusing because everytime i need to know what is the first, second, or third usize are I have to read the function code. I can't even tell what is the last usize is even after reading the code (what to call `responses_expected - (err_count + success_count)` is that the messages that was never received ?
You have 2 options 
- 1) (preferred) return a structure: so fields will be named hence I can look at the struct and know which field I want to check.
- 2) create type aliased for each "Type" you return for example
```rust
type ErrorCount = usize
type SuccessCount = usize
type LostCount = usize

// then you function can return
fn test() -> (Vec<Message>, ErrorCount, SuccessCount, LostCount)
```

Reply:
I admit the code here looks ugly. will try to improve the readability here as you suggested. and yes, as you guessed, it is the count of the messages that were lost/never received.